### PR TITLE
Python Windows fixes (ready)

### PR DESCRIPTION
**Pull request by rcurtin**
This is collecting fixes for #1872.  @gmanlan let me know if these fixes solve the problems you're having on Windows.  As we get to the bottom of more issues we can add to this PR and merge when we have it working.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,7 +20,9 @@ if (WIN32)
   option(BUILD_PYTHON_BINDINGS "Build Python bindings." OFF)
   option(BUILD_SHARED_LIBS
       "Compile shared libraries (if OFF, static libraries are compiled)." OFF)
-  message(WARNING "By default Python bindings are not compiled for Windows because they are not known to work.  Set BUILD_PYTHON_BINDINGS to ON if you want them built.")
+  if (NOT BUILD_PYTHON_BINDINGS)
+    message(WARNING "By default Python bindings are not compiled for Windows because they are not known to work.  Set BUILD_PYTHON_BINDINGS to ON if you want them built.")
+  endif ()
```
Comment (gmanlan) [inline CMakeLists.txt]: looks good

**Comment by gmanlan** [comment]
Keep Alive

**Comment by rcurtin** [comment]
Okay, not quite sure what I did with the Windows git client to push like that, but I'll figure it out tomorrow... `21dc0f7` is the commit that should fix the runtime DLL issue.

**Comment by rcurtin** [comment]
Ok, fixed the commit issue; hopefully this will pass the build tests now---we'll see. :)

**Comment by rcurtin** [comment]
Alright, I think that this is ready for review.  There may still be a few small build issues that I have to work out, but both @gmanlan and I have tested this and had success.  Once this is merged, I can then start making the mlpack/mlpack-wheels repository build on Windows too, so that we can have mlpack PyPI packages for Windows also.

File: CMakeLists.txt (hunk 1 above)
Comment (gmanlan) [reply CMakeLists.txt]: (^)

File: CMakeLists.txt [hunk 2]
```diff
@@ -21,7 +21,12 @@ if (WIN32)
   option(BUILD_PYTHON_BINDINGS "Build Python bindings." OFF)
   option(BUILD_SHARED_LIBS
       "Compile shared libraries (if OFF, static libraries are compiled)." OFF)
-  message(WARNING "By default Python bindings are not compiled for Windows because they are not known to work.  Set BUILD_PYTHON_BINDINGS to ON if you want them built.")
+  if (NOT BUILD_PYTHON_BINDINGS)
+    message(WARNING "By default Python bindings are not compiled for Windows because they are not known to work.  Set BUILD_PYTHON_BINDINGS to ON if you want them built.")
```
Comment (gmanlan) [inline CMakeLists.txt]: I think this warning should be deprecated right?

Comment (rcurtin) [reply CMakeLists.txt]: Good point, thanks!  Let me fix that.

**Comment by gmanlan** [comment]
Updates on this one..?

**Comment by rcurtin** [comment]
Just waiting on a review.

@zoq @ShikharJ @KimSangYeon-DGU @walragatver (or anyone else), any interest in reviewing this one?  It should be good to go, just needs some eyes on it to see if anything can be improved in the code. :)  (No huge hurry, but I would love to get it into mlpack 3.2.2.)

**Comment by zoq** [comment]
Sorry for the slow response, thought this is still WIP (see title).

**Comment by rcurtin** [comment]
Thanks---I didn't realize that I had the `(WIP)` in the name still (I updated it).  It definitely took a long time to make it all work but I think it was worth it. :)  I wish I could have reviewed code on the plane too but the wifi didn't work for me and I wasn't smart enough to preload some PRs in tabs to review :angry:

**Pull request merged**